lib/puppet/pops/merge_strategy.rb

```diff
@@ -0,0 +1,287 @@
+    end
+
+    def assert_type(param, type, value)
+      TypeAsserter.assert_instance_of(param, type, value)
+    end
+  end
+
+  # Produces a new hash by merging hash e1 with hash e2 in such a way that the values of duplicate keys
+  # will be those of e1
+  #
+  class HashMergeStrategy < MergeStrategy
+    def self.key
+      :hash
+    end
+
+    # @param e1 [Hash<String,Object>] The hash that will act as the source of the merge
+    # @param e2 [Hash<String,Object>] The hash that will act as the receiver for the merge
+    # @return [Hash<String,Object]] The merged hash
+    # @see Hash#merge
+    def checked_merge(e1, e2)
+      e2.merge(e1)
+    end
+
+    protected
+
+    def value_t
+      @value_t ||= Puppet::Pops::Types::TypeParser.new.parse('Hash[String,Data]')
+    end
+
+    MergeStrategy.add_strategy(self)
+  end
+
+  # Merges two values that must be either scalar or arrays into a unique set of values.
+  #
+  # Scalar values will be converted into a one element arrays and array values will be flattened
+  # prior to forming the unique set. The order of the elements is preserved with e1 being the
+  # first contributor of elements and e2 the second.
+  #
+  class UniqueMergeStrategy < MergeStrategy
+    def self.key
+      :unique
+    end
+
+    # @param e1 [Array<Object>] The first array
+    # @param e2 [Array<Object>] The second array
+    # @return [Array<Object>] The unique set of elements
+    #
+    def checked_merge(e1, e2)
+      convert_value(e1) | convert_value(e2)
+    end
+
+    def convert_value(e)
+      e.is_a?(Array) ? e.flatten : [e]
+    end
+
+    protected
+
+    def value_t
+      @value_t ||= Puppet::Pops::Types::TypeParser.new.parse('Variant[Scalar,Array[Data]]')
+    end
+
+    MergeStrategy.add_strategy(self)
+  end
+
+  # Documentation copied from https://github.com/danielsdeleo/deep_merge/blob/master/lib/deep_merge/core.rb
+  # altered with respect to _preserve_unmergeables_ since this implementation always disables that option.
+  #
+  # The destination is dup'ed before the deep_merge is called to allow frozen objects as values.
+  #
+  # deep_merge method permits merging of arbitrary child elements. The two top level
+  # elements must be hashes. These hashes can contain unlimited (to stack limit) levels
+  # of child elements. These child elements to not have to be of the same types.
+  # Where child elements are of the same type, deep_merge will attempt to merge them together.
+  # Where child elements are not of the same type, deep_merge will skip or optionally overwrite
+  # the destination element with the contents of the source element at that level.
+  # So if you have two hashes like this:
+  #   source = {:x => [1,2,3], :y => 2}
+  #   dest =   {:x => [4,5,'6'], :y => [7,8,9]}
+  #   dest.deep_merge!(source)
+  #   Results: {:x => [1,2,3,4,5,'6'], :y => 2}
+  #
+  # "deep_merge" will unconditionally overwrite any unmergeables and merge everything else.
+  #
+  # Options:
```

Comment:
I suspect that the layout of the options section is not going to be what is expected